test/e2e/lb_subnets_test.go

```diff
@@ -0,0 +1,419 @@
+	}
+	t.Cleanup(func() { assertIngressControllerDeleted(t, kclient, ic) })
+
+	// Wait for the load balancer and DNS to be ready.
+	if err := waitForIngressControllerCondition(t, kclient, 10*time.Minute, name, availableConditionsForIngressControllerWithLoadBalancer...); err != nil {
+		t.Fatalf("failed to observe expected conditions: %v", err)
+	}
+
+	// Verify the Service annotation is unset by default.
+	lbService := &v12.Service{}
+	if err := kclient.Get(context.Background(), controller.LoadBalancerServiceName(ic), lbService); err != nil {
+		t.Fatalf("failed to get lb service: %v", err)
+	}
+	if val, ok := lbService.Annotations[awsLBSubnetAnnotation]; ok {
+		t.Fatalf("unexpected %q annotation: %q", awsLBSubnetAnnotation, val)
+	}
+
+	// Make sure the IngressController with no specified subnets has connectivity.
+	externalTestPodName := types.NamespacedName{Name: name.Name + "-external-verify", Namespace: name.Namespace}
+	testHostname := "apps." + ic.Spec.Domain
+	t.Logf("verifying ingresscontroller %q using CLB with no specified subnets has external connectivity", ic.Name)
+	if err := verifyExternalIngressController(t, externalTestPodName, testHostname, testHostname, 10*time.Minute); err != nil {
+		t.Fatalf("failed to verify connectivity with workload with hostname %s using external client: %v", testHostname, err)
+	}
+
+	// Update the IngressController to add the public subnets we discovered.
+	icName := types.NamespacedName{Name: ic.Name, Namespace: ic.Namespace}
+	t.Logf("updating ingresscontroller %q to use public subnets", ic.Name)
+	if err := updateIngressControllerWithRetryOnConflict(t, icName, 5*time.Minute, func(ic *operatorv1.IngressController) {
+		ic.Spec.EndpointPublishingStrategy.LoadBalancer.ProviderParameters = &operatorv1.ProviderLoadBalancerParameters{
+			Type: operatorv1.AWSLoadBalancerProvider,
+			AWS: &operatorv1.AWSLoadBalancerParameters{
+				Type: operatorv1.AWSClassicLoadBalancer,
+				ClassicLoadBalancerParameters: &operatorv1.AWSClassicLoadBalancerParameters{
+					Subnets: publicSubnets,
+				},
+			},
+		}
+	}); err != nil {
+		t.Fatalf("failed to update ingresscontroller: %v", err)
+	}
+
+	// Effectuate the updated IngressController subnets.
+	if err := effectuateIngressControllerSubnets(t, ic, publicSubnets); err != nil {
+		t.Fatalf("failed to effectuate the ingresscontroller subnets: %v", err)
+	}
+
+	// Verify we can still reach the CLB with the provided public subnets.
+	t.Logf("verifying external connectivity for ingresscontroller %q using an CLB with specified public subnets", ic.Name)
+	if err := verifyExternalIngressController(t, externalTestPodName, testHostname, testHostname, 10*time.Minute); err != nil {
+		t.Fatalf("failed to verify connectivity with workload with hostname %s using external client: %v", testHostname, err)
+	}
+
+	// Now, update the IngressController to change to use the private subnets, so it can't connect externally.
```

Comment:
Does it test something additionally to what has already tested when we passed from "IngressController with no subnets" to "IngressController with public subnets"? To me it seems to be the same code path, no?

I tried to think of a potential way to shorten this e2e. Would the below cover all the cases?

- Create IngressController with public subnets, wait for the LB to be provisioned, verifyExternal
- Update the IngressController from CLB to NLB removing 1 public subnet (I suppose we still will have `>=1`), verifyExternal
- Remove subnets with auto delete and verify the annotation is removed (just as now)

Also, these scenarios will test the update of the subnets without the need to wait for "not connect" to private subnets. Therefore `verifyExternalIngressController` can be left unchanged.